imgui/src/drag_drop.rs

```diff
@@ -0,0 +1,475 @@
+/// }
+/// ```
+///
+/// Notice especially the `"BUTTON_DRAG"` and `"BUTTON_ID"` name -- this is the identifier of this
+/// DragDropTarget; [DragDropSource]s will specify an identifier when they send a payload, and these
+/// names must match up. Notice how a target can have multiple acceptances on them -- this is a good
+/// pattern to handle multiple kinds of data which could be passed around.
+///
+/// DropDropTargets don't do anything until you use one of the three `accept_` methods
+/// on this struct. Each of these methods will spit out a _Payload struct with an increasing
+/// amount of information on the Payload. The absolute safest solution is [accept_payload_empty](Self::accept_payload_empty).
+#[derive(Debug)]
+pub struct DragDropTarget<'ui>(PhantomData<Ui<'ui>>);
+
+impl<'ui> DragDropTarget<'ui> {
+    /// Creates a new DragDropTarget, holding the [Ui]'s lifetime for the duration
+    /// of its existence. This is required since this struct runs some code on its Drop
+    /// to end the DragDropTarget code.
+    pub fn new(_ui: &Ui<'_>) -> Option<Self> {
+        let should_begin = unsafe { sys::igBeginDragDropTarget() };
+        if should_begin {
+            Some(Self(PhantomData))
+        } else {
+            None
+        }
+    }
+
+    /// Accepts an empty payload. This is the safest option for raising named events
+    /// in the DragDrop API. See [DragDropSource::begin] for more information on how you
+    /// might use this pattern.
+    pub fn accept_payload_empty(
+        &self,
+        name: &ImStr,
+        flags: DragDropFlags,
+    ) -> Option<DragDropPayloadEmpty> {
+        self.accept_payload::<()>(name, flags)?
+            .ok()
+            .map(|payload_pod| DragDropPayloadEmpty {
+                preview: payload_pod.preview,
+                delivery: payload_pod.delivery,
+            })
+    }
+
+    /// Accepts a payload with plain old data in it. This returns a Result, since you can specify any
+    /// type. The sent type must match the return type (via TypeId) to receive an `Ok`.
+    pub fn accept_payload<T: 'static + Copy>(
+        &self,
+        name: &ImStr,
+        flags: DragDropFlags,
+    ) -> Option<Result<DragDropPayloadPod<T>, PayloadIsWrongType>> {
+        let output = unsafe { self.accept_payload_unchecked(name, flags) };
+
+        // convert the unsafe payload to our Result
+        output.map(|unsafe_payload| {
+            // sheering off the typeid...
+            let received = unsafe { *(unsafe_payload.data as *const any::TypeId) };
+            let requested = any::TypeId::of::<T>();
+
+            if received == requested {
+                let data = unsafe { *(unsafe_payload.data as *const TypedPayload<T>) }.data;
+                Ok(DragDropPayloadPod {
+                    data,
+                    preview: unsafe_payload.preview,
+                    delivery: unsafe_payload.delivery,
+                })
+            } else {
+                Err(PayloadIsWrongType {
```

Comment:
Certainly here we could easily know [`type_name`](https://doc.rust-lang.org/std/any/fn.type_name.html) of the expected type. If we store the type_name of the actual type on TypedPayload, the debugging experience might improve as mentioned elsewhere.